src/Tools/Source/RunTests/Cache/ContentUtil.cs

```diff
@@ -206,17 +215,34 @@ private static string HashBytesToString(byte[] hash)
 
         private string GetFileChecksum(string filePath)
         {
-            string checksum;
-            if (_fileToChecksumMap.TryGetValue(filePath, out checksum))
+            var ext = Path.GetExtension(filePath).ToLower();
+            return (ext == ".dll" || ext == ".exe") 
+                ? GetAssemblyChecksum(filePath)
+                : GetFileChecksumCore(filePath);
+        }
+
+        private string GetAssemblyChecksum(string filePath)
+        {
+            var mvid = GetAssemblyMvid(filePath);
+            if (!_assemblyChecksumCacheMap.TryGetValue(mvid, out var checksum))
             {
-                return checksum;
+                checksum = GetFileChecksumCore(filePath);
+                _assemblyChecksumCacheMap[mvid] = checksum;
             }
 
-            checksum = GetFileChecksumCore(filePath);
-            _fileToChecksumMap.Add(filePath, checksum);
             return checksum;
         }
 
+        private Guid GetAssemblyMvid(string filePath)
+        {
+            using (var source = File.Open(filePath, FileMode.Open, FileAccess.Read, FileShare.Read))
+            using (var reader = new PEReader(source))
+            {
+                var metadataReader = reader.GetMetadataReader();
+                return metadataReader.GetGuid(metadataReader.GetModuleDefinition().Mvid);
+            }
+        }
+
         private string GetFileChecksumCore(string filePath)
         {
             var bytes = File.ReadAllBytes(filePath);
```

Comment:
Could [`HashAlgorithm.ComputeHash(Stream)`](https://msdn.microsoft.com/en-us/library/xa627k19(v=vs.110).aspx) be used instead of reading file contents into byte arrays?